### PR TITLE
Airfoil Parameters surrogate model example

### PR DESCRIPTION
**Pull request by raul-rufato**
I am implementing an example of a surrogate model that uses SMT to be able to determine the aerodinqmic parameters of a given aerofoil, such as cd,cl,cm. And it uses the mode shapes, Mach number, angle of attack and reynolds number as inputs. I believe the best place to put this example is in this section, if not please tell me. The example was implemented in a .py file with all comments and all the necessary files are attached.

### CODE REVIEW

File: smt/examples/airfoil_parameters/learning_airfoils_parameters.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from smt.surrogate_models.genn import GENN, load_smt_data
+import csv
+from sklearn.model_selection import train_test_split 
+import pickle
+import tensorflow as tf
```
Comment (relf) [inline smt/examples/airfoil_parameters/learning_airfoils_parameters.py:7]: remove unused import

File: smt/examples/airfoil_parameters/learning_airfoils_parameters.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+import matplotlib.pyplot as plt
```
Comment (relf) [inline smt/examples/airfoil_parameters/learning_airfoils_parameters.py:8]: duplicated import

File: smt/examples/airfoil_parameters/learning_airfoils_parameters.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+import pandas as pd
```
Comment (relf) [inline smt/examples/airfoil_parameters/learning_airfoils_parameters.py:9]: panda is used to read csv, just use csv and remove panda dependency

File: smt/examples/airfoil_parameters/learning_airfoils_parameters.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+"""
+Predicting Airfoil Aerodynamics through data
+by Raul Carreira Rufato and Prof. Joseph Morlier
+
+This notebook is a tutorial to determine the aerodynamic coefficients of a given airfoil using SMT. The obtained surrogate model can be used to give predictions for certain Mach numbers, angles of attack and the aerodynamic coefficients. These calculations can be really usefull in case of an airfoil shape optimization. The input parameters uses the airfoil Camber and Thickness mode shapes.
```
Comment (relf) [inline smt/examples/airfoil_parameters/learning_airfoils_parameters.py:15]: format text to have 80 to 100 characters lines

File: smt/examples/airfoil_parameters/learning_airfoils_parameters.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
```
Comment (relf) [inline smt/examples/airfoil_parameters/learning_airfoils_parameters.py:16]: remove blank lines

File: smt/examples/airfoil_parameters/learning_airfoils_parameters.py (hunk 4 above)
Comment (relf) [reply smt/examples/airfoil_parameters/learning_airfoils_parameters.py:15]: not a notebook anymore  ;-)

File: smt/examples/airfoil_parameters/learning_airfoils_parameters.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+Inputs: Airfoil Camber and Thickness mode shapes, Mach, alpha
+
+Outputs (options): cd, cl, cm
+
+In this test case, we will be predicting only the Cd coefficient. However, the other databases for the prediction of the other terms are available in the same repository.
+
+Bouhlels mSANN in a nutshell:
+
+a. Using inverse distance weighting (IDW) to interpolate the surface function of each airfoil.
```
Comment (relf) [inline smt/examples/airfoil_parameters/learning_airfoils_parameters.py:25]: below `GENN` is used, not `IDW`? Maybe add a short explicit description of what is `mSANN` exactly, there are 4 steps below but do not understand how they relate to it. mSANN, what is it and what is it for?

**Review by relf — CHANGES_REQUESTED** [review]
You have to pass `black` to format the script. Is there a special reason to use GENN?

**Comment by raul-rufato** [comment]
Hello, I've made the suggested changes. Let me know if it works for you.

Kind regards, 

Raul

**Comment by raul-rufato** [comment]
Hi,
Thanks for your response. I used GENN in the example as I am reproducing a work that used the same, so I prefered not to change it. 

I've ran black on the .py file. Thanks for the tip, I didn't know it before. 

When all checks are finished, let me know if it is better for you now. 

Regards, 

Raul

File: smt/examples/airfoil_parameters/learning_airfoils_parameters.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from smt.surrogate_models.genn import GENN, load_smt_data
+import csv
+from sklearn.model_selection import train_test_split
+
+"""
+Predicting Airfoil Aerodynamics through data by Raul Carreira Rufato and Prof. Joseph Morlier
+This is a tutorial to determine the aerodynamic coefficients of a given airfoil using GENN in SMT (Other models could be used as well)
```
Comment (relf) [inline smt/examples/airfoil_parameters/learning_airfoils_parameters.py:9]: Other -> other

**Review by relf — CHANGES_REQUESTED** [review]
Is it mSANN you are reproducing?

Please answer the review by fill in the textfield below each comment or question, it is easier to validate your answers.

File: smt/examples/airfoil_parameters/learning_airfoils_parameters.py (hunk 7 above)
Comment (raul-rufato) [reply smt/examples/airfoil_parameters/learning_airfoils_parameters.py:9]: """
Predicting Airfoil Aerodynamics through data by Raul Carreira Rufato and Prof. Joseph Morlier
This is a tutorial to determine the aerodynamic coefficients of a given airfoil using GENN in SMT (other models could be used as well)

**Comment by raul-rufato** [comment]
> Is it mSANN you are reproducing?
> 
> Please answer the review by filling in the textfield below each comment or question, it is easier to validate your answers.

```
"""
Predicting Airfoil Aerodynamics through data by Raul Carreira Rufato and Prof. Joseph Morlier
This is a tutorial to determine the aerodynamic coefficients of a given airfoil using GENN in SMT (other models could be used as well) 
The obtained surrogate model can be used to give predictions for certain Mach numbers, angles of attack and the aerodynamic coefficients. 
These calculations can be really useful in case of an airfoil shape optimization. The input parameters uses the airfoil Camber and Thickness mode shapes.
Inputs: Airfoil Camber and Thickness mode shapes, Mach, alpha
Outputs (options): cd, cl, cm
In this test case, we will be predicting only the Cd coefficient. However, the other databases for the prediction of the 
other terms are available in the same repository. Bouhlels mSANN uses the information contained in the paper [4] to determine 
the airfoil's mode shapes. Moreover, in mSANN a deep neural network is used to predict the Cd parameter of a given parametrized
airfoil. Therefore, in this tutorial, we reproduce the paper [1] using the Gradient-Enhanced Neural Networks (GENN)  from SMT. 
Briefly explaining how mSANN generates the mode shapes of a given airfoil:
a. Using inverse distance weighting (IDW) to interpolate the surface function of each airfoil.
b. Then applying singular value decomposition (SVD) to reduce the number of variables that define the airfoil geometry. It includes a total of 
14 airfoil modes (seven for camber and seven for thickness).
c. Totally 16 input variables, two flow conditions of Mach number (0.3 to 0.6) and the angle of attack (2 degrees to 6 degrees) plus 14 shape coefficients.
d. The output airfoil aerodynamic force coefficients and their respective gradients are computed using ADflow, which solves the RANS equations with a 
Spalart-Allmaras turbulence model.
[1] Bouhlel, M. A., He, S., and Martins, J. R. R. A., mSANN Model Benchmarks, Mendeley Data, 2019. https://doi.org/10. 17632/ngpd634smf.1.
[2] Li, J., Bouhlel,M., and Martins,J.R.R.A., Data-Based Approach for Fast Airfoil Analysis. and Optimization,AIAA Journal, 2019.
[3] Bouhlel, M.A., and Martins,J.R.R.A.,Gradient-EnhancedKrigingforHigh-DimensionalProblems,Springer-Verlag,2018.
[4] Bouhlel,M., He,S., and Martins,J.R.R.A.,Scalable Gradient-Enhanced Artificial Neural Networks for Airfoil Shape Design in the Subsonic and Transonic 
Regimes, ResearchGate, 2020.
[5] Du,X.,He,P.,and Martins,J.,Rapid airfoil design optimization via neuralnetworks-basedp arameterization and surrogate modeling, Elsevier, 2021.
[6] University of Michigan, Webfoil, 2021. URL http://webfoil.engin.umich.edu/, online accessed on 16 of June 2021.
"""
```

**Comment by raul-rufato** [comment]
Should I re do the pull request after change the comments?
I’am explaining it better here:

"""
Predicting Airfoil Aerodynamics through data by Raul Carreira Rufato and Prof. Joseph Morlier
This is a tutorial to determine the aerodynamic coefficients of a given airfoil using GENN in SMT (other models could be used as well)
The obtained surrogate model can be used to give predictions for certain Mach numbers, angles of attack and the aerodynamic coefficients.
These calculations can be really useful in case of an airfoil shape optimization. The input parameters uses the airfoil Camber and Thickness mode shapes.
Inputs: Airfoil Camber and Thickness mode shapes, Mach, alpha
Outputs (options): cd, cl, cm
In this test case, we will be predicting only the Cd coefficient. However, the other databases for the prediction of the
other terms are available in the same repository. Bouhlels mSANN uses the information contained in the paper [4] to determine
the airfoil's mode shapes. Moreover, in mSANN a deep neural network is used to predict the Cd parameter of a given parametrized
airfoil. Therefore, in this tutorial, we reproduce the paper [1] using the Gradient-Enhanced Neural Networks (GENN)  from SMT.
Briefly explaining how mSANN generates the mode shapes of a given airfoil:
a. Using inverse distance weighting (IDW) to interpolate the surface function of each airfoil.
b. Then applying singular value decomposition (SVD) to reduce the number of variables that define the airfoil geometry. It includes a total of
14 airfoil modes (seven for camber and seven for thickness).
c. Totally 16 input variables, two flow conditions of Mach number (0.3 to 0.6) and the angle of attack (2 degrees to 6 degrees) plus 14 shape coefficients.
d. The output airfoil aerodynamic force coefficients and their respective gradients are computed using ADflow, which solves the RANS equations with a
Spalart-Allmaras turbulence model.
[1] Bouhlel, M. A., He, S., and Martins, J. R. R. A., mSANN Model Benchmarks, Mendeley Data, 2019. https://doi.org/10. 17632/ngpd634smf.1.
[2] Li, J., Bouhlel,M., and Martins,J.R.R.A., Data-Based Approach for Fast Airfoil Analysis. and Optimization,AIAA Journal, 2019.
[3] Bouhlel, M.A., and Martins,J.R.R.A.,Gradient-EnhancedKrigingforHigh-DimensionalProblems,Springer-Verlag,2018.
[4] Bouhlel,M., He,S., and Martins,J.R.R.A.,Scalable Gradient-Enhanced Artificial Neural Networks for Airfoil Shape Design in the Subsonic and Transonic
Regimes, ResearchGate, 2020.
[5] Du,X.,He,P.,and Martins,J.,Rapid airfoil design optimization via neuralnetworks-basedp arameterization and surrogate modeling, Elsevier, 2021.
[6] University of Michigan, Webfoil, 2021. URL http://webfoil.engin.umich.edu/, online accessed on 16 of June 2021.
"""

**Comment by relf** [comment]
I was just saying to explain your changes not copying them in the texfield. For instance for fixing `usefull`, trivially just say `done` and resolve the conversation. If I ask 'Is it mSANN' you are trying to reproduce?', answer 'yes' and maybe tell me that you change the text accordingly. 

Thus,
> Moreover, in mSANN a deep neural network is used to predict the Cd parameter of a given parametrized
airfoil. Therefore, in this tutorial, we reproduce the paper [1] using the Gradient-Enhanced Neural Networks (GENN)  from SMT.

was the answer to my question regarding why you want to use GENN, (I had to dive in your answer to find it, just keep to the point).

No need to do another pull request! Just commit and push your changes. 
No need to copy mails you received, it just clutters the conversation.

**Comment by jomorlier** [comment]
Are Raul changes OK ?

File: smt/examples/airfoil_parameters/learning_airfoils_parameters.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        dim_values = values.shape
+        x = values[:, : dim_values[1] - 1]
+        y = values[:, -1]
+    with open("DataDySMTCd.csv") as file:
+        reader = csv.reader(file, delimiter=";")
+        dy = np.array(list(reader), dtype=np.float32)
+    return x, y, dy
+
+
+def graphPredictionsSMT(airfoil_modeshapes, airfoil_name, Ma, plot: bool, genn):
```
Comment (relf) [inline smt/examples/airfoil_parameters/learning_airfoils_parameters.py:48]: This method is defined twice!

**Review by relf — CHANGES_REQUESTED** [review]
@raul-rufato To make your example visible in the documentation, you have to follow [doc build instructions](https://smt.readthedocs.io/en/latest/_src_docs/dev_docs.html#building-the-documentation) and see what is done to make it work in other examples [here](https://github.com/SMTorg/smt/blob/1fa376c6e83be57d742470619e205933cb69a4db/doc/_src_docs/examples.rstx) and [there](https://github.com/SMTorg/smt/tree/master/doc/_src_docs/examples). Ensure you can generate the doc locally before commiting your changes.

This is not quite simple... If you want I could take over your PR to do it later on.

File: smt/examples/airfoil_parameters/learning_airfoils_parameters.py (hunk 8 above)
Comment (raul-rufato) [reply smt/examples/airfoil_parameters/learning_airfoils_parameters.py]: Done

File: smt/examples/airfoil_parameters/learning_airfoils_parameters.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+        alpha[i, 0] = -1 + 0.16 * i
+    input_array = np.concatenate((new_input_array, alpha), axis=1)
+    # predictions are made
+    cd_pred = modelcd.predict_values(input_array)
+    with open("NACA4412-ADflow-alpha-cd.csv") as file:
+        reader = csv.reader(file, delimiter=" ")
+        cd_adflow = np.array(list(reader)[1:], dtype=np.float32)
+    # cd from ADflow and Xfoil
+    # graphs for the single aerodynamic coefficients are computed -> through bool: plot it is to decide if graphs are computed or not
+    if plot == True:
```
[the remark this answers is not in the document]
Comment (raul-rufato) [reply smt/examples/airfoil_parameters/learning_airfoils_parameters.py]: Done

**Comment by raul-rufato** [comment]
@relf Thank you for the new suggestions.

 I do prefer you to take over my PR to do it later on, thank you again.

**Comment by jomorlier** [comment]
Just only ref 1 &2 are hyperlinks


Le Vendredi, Juillet 01, 2022 16:17 CEST, Rémi Lafage ***@***.***> a écrit:

**Comment by relf** [comment]
@jomorlier, yes because in the initial text only ref 1 and 2 were actually referenced
